### PR TITLE
cnf ran: ztp hub templating and node delete migration

### PR DESCRIPTION
**Pull request by klaskosk**

### CODE REVIEW

**Comment by klaskosk** [comment]
Updated to turn the hub templating tests into positive tests and restrict them to TALM >=4.16

File: tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+			By("checking the TALM version")
+			versionInRange, err := ranhelper.IsVersionStringInRange(RANConfig.HubOperatorVersions[ranparam.TALM], "4.16", "")
+			Expect(err).ToNot(HaveOccurred(), "Failed to check if TALM version is in range")
+
+			if !versionInRange {
+				Skip("This test requires a TALM version of at least 4.16")
+			}
+		})
+
+		// 54240 - Hub-side ACM templating with TALM
```
Comment (klaskosk) [inline tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go:87]: I've already created the Jira for getting these test IDs changed

File: tests/cnf/ran/ztp/internal/helper/nodedelete.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	"github.com/openshift-kni/eco-goinfra/pkg/nodes"
+	. "github.com/openshift-kni/eco-gotests/tests/cnf/ran/internal/raninittools"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/ran/ztp/internal/tsparams"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// IsSnoPlusOne checks if the specified cluster has one master and one worker node.
+func IsSnoPlusOne(client *clients.Settings) (bool, error) {
+	masters, err := listNodesByRole(client, "master")
```
Comment (sebrandon1) [inline tests/cnf/ran/ztp/internal/helper/nodedelete.go:21]: As of Kubernetes 1.24, the `master` label is no longer applied to nodes:
https://kubernetes.io/blog/2022/04/07/upcoming-changes-in-kubernetes-1-24/#api-removals-deprecations-and-other-changes-for-kubernetes-1-24

`node-role.kubernetes.io/master` changes to `node-role.kubernetes.io/control-plane`.  In the past I've just checked for both `master` and `control-plane` to keep backwards compatibility.

Similar to: https://github.com/openshift-kni/eco-goinfra/pull/265

Comment (sebrandon1) [reply tests/cnf/ran/ztp/internal/helper/nodedelete.go:21]: We can potentially do this in a separate PR too to adjust this for the rest of the repo.

**Review by sebrandon1 — COMMENTED** [review]
One comment about the master/control-plane stuff.

File: tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+			By("validating the policy reaches compliant status")
+			policy, err := ocm.PullPolicy(HubAPIClient, tsparams.HubTemplatingPolicyName, tsparams.TestNamespace)
+			Expect(err).ToNot(HaveOccurred(), "Failed to get policy from hub cluster")
+
+			err = policy.WaitUntilComplianceState(policiesv1.Compliant, tsparams.ArgoCdChangeTimeout)
+			Expect(err).ToNot(HaveOccurred(), "Failed to wait for policy to become compliant")
+		})
+
+		// 54240 - Hub-side ACM templating with TALM
+		It("should allow templating with the prinf function", reportxml.ID("54240"), func() {
```
Comment (yliu127) [inline tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go:95]: description does not match test content. And it is now a valid test case. so should be moved to	`When("supported ACM hub side templating is used", func() {`

File: tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+			By("validating the policy reaches compliant status")
+			policy, err := ocm.PullPolicy(HubAPIClient, tsparams.HubTemplatingPolicyName, tsparams.TestNamespace)
+			Expect(err).ToNot(HaveOccurred(), "Failed to get policy from hub cluster")
+
+			err = policy.WaitUntilComplianceState(policiesv1.Compliant, tsparams.ArgoCdChangeTimeout)
+			Expect(err).ToNot(HaveOccurred(), "Failed to wait for policy to become compliant")
+		})
+
+		// 54240 - Hub-side ACM templating with TALM
+		It("should report an error for using autoindent function where not allowed", reportxml.ID("54240"), func() {
```
Comment (yliu127) [inline tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go:113]: I don't think this test is new to 4.16. This was introduced in 4.12.

File: tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+			By("validating the specific error using the policy message")
+			err := helper.WaitForPolicyMessageToContainSubstring(
+				HubAPIClient,
+				tsparams.TestNamespace+"."+tsparams.HubTemplatingPolicyName,
+				RANConfig.Spoke1Name,
+				"wrong type for value; expected string; got int")
+			Expect(err).ToNot(HaveOccurred(), "Failed to validate error using policy message")
+		})
+
+		It("should allow a previously invalid templating function", func() {
```
Comment (yliu127) [inline tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go:128]: This test case can be removed, and a new valid template should be added to include the new valid configs. 

This should be created with existing content: 
http://registry.kni-qe-0.lab.eng.rdu2.redhat.com:3000/kni-qe/ztp-site-configs/src/workers-4.16/policygentemplates/ztp-test/hub-templating-valid-4.15

And the hub-templating-valid should hold the latest starting 4.16.

When the test cases were created, the idea was negative test cases require one template for each test case in order to validate the proper error message. However, valid templates should be consolidated into single test to reduce execution time. This approach will make it easier to add new valid scenarios. i.e., minimum eco-gotest change is needed. 

This is an example of how different valid test cases are handled between 4.11 and 4.12+: 
https://gitlab.cee.redhat.com/cnf/cnf-gotests/-/blob/release-4.14/test/ran/ztp/tests/ztp_argocd_hub_templating.go?ref_type=heads#L205-212

Sorry I forgot about the history when I review the cnf-gotests PR until I see the big picture in this PR. 

CC @josclark42

File: tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+			versionInRange, err := ranhelper.IsVersionStringInRange(RANConfig.HubOperatorVersions[ranparam.TALM], "4.16", "")
+			Expect(err).ToNot(HaveOccurred(), "Failed to check if TALM version is in range")
+
+			if !versionInRange {
+				Skip("This test requires a TALM version of at least 4.16")
+			}
+		})
+
+		// 54240 - Hub-side ACM templating with TALM
+		It("should allow templating with the printf function", reportxml.ID("54240"), func() {
```
Comment (yliu127) [inline tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go:83]: This is now a positive test case, so it should be merged into a single valid test case and moved to this  container: 
`"	When("supported ACM hub side templating is used", func() {"`

File: tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	"github.com/openshift-kni/eco-gotests/tests/cnf/ran/internal/ranparam"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/ran/ztp/internal/helper"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/ran/ztp/internal/tsparams"
+	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+)
+
+var _ = Describe("ZTP Argo CD Hub Templating Tests", Label(tsparams.LabelArgoCdHubTemplatingTestCases), func() {
+	BeforeEach(func() {
+		By("checking the ZTP version")
+		versionInRange, err := ranhelper.IsVersionStringInRange(RANConfig.ZTPVersion, "4.11", "")
```
Comment (yliu127) [inline tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go:28]: I think we can change this to 4.12, as the negative test case was introduced in 4.12.  Plus we no longer run 4.11 CI.

Comment (klaskosk) [reply tests/cnf/ran/ztp/tests/ztp-argocd-hub-templating.go:28]: I changed the message but not the code oops, I'll fix that

**Review by yliu127 — APPROVED** [review]

**Review by sebrandon1 — APPROVED** [review]

**Pull request merged**